daemon/cmd/endpoint.go

```diff
@@ -137,11 +148,19 @@ func NewGetEndpointIDHandler(d *Daemon) GetEndpointIDHandler {
 func (h *getEndpointID) Handle(params GetEndpointIDParams) middleware.Responder {
 	log.WithField(logfields.EndpointID, params.ID).Debug("GET /endpoint/{id} request")
 
+	r, err := h.d.apiLimiterSet.Wait(params.HTTPRequest.Context(), apiRequestEndpointGet)
+	if err != nil {
+		return api.Error(responseTooManyRequests, err)
```

Comment:
@tgraf IIUC there are primarily two paths for a failed API request: 1. the request exceeds the maximum wait time, 2. the request exceeds the rate limiting settings.  The latter will return immediately with a return code of *responseTooManyRequests* (42), whereas the former will return with another failure.

My questions, as a user are: does Cilium somehow update the status on the pod to reflect the cause of the error?  Secondly, since rate limiting is intended as a means to reduce pressure on the cilium agent, what is the recommended course of action for the user upon receiving *responseTooManyRequests*?  Again, my understanding is that the user has two options: either to retry after a short amount of time, or to adjust the rate-limiting configuration and then retry immediately.  My issue with the second approach is that the end user--the one who initiated the pod creation--typically won't have the ability to make cluster wide changes.  Hence, my question is, what is the mode for interrogation and the passing of these failures and settings across the multiple components (end-user, *k8s* operator, cilium)?

Reply:
> does Cilium somehow update the status on the pod to reflect the cause of the error?

The error will surface in `kubectl describe pod` and `kubectl get events` so you will see that CNI creation failed because of too many requests. It should give you better visibility than before instead of kubelet/runtime timing out and retrying silently.

> Secondly, since rate limiting is intended as a means to reduce pressure on the cilium agent, what is the recommended course of action for the user upon receiving responseTooManyRequests?

In general it should resolve itself automatically as kubelet will continue to retry. It will avoid the dangerous CNI ADD / CNI DEL cycle after which some resources have already been invested and then have to be undone.

> Again, my understanding is that the user has two options: either to retry after a short amount of time,

kubelet will take care of this. The difference will be that it is done with visibility and it fails upfront without an investment of CPU resources already committed.

> or to adjust the rate-limiting configuration and then retry immediately. 

Adjustment of rate-limiting configuration should only be needed if the automatic adjustment doesn't allow for the required number of pod creations (or other API calls) even if low processing durations are observed. If enough resources are available, the default rate limits will automatically be relaxed further. An example where things could not work out well automatically: if an API call is estimated to take 2 seconds with 4 parallel requests but it only takes 0.5 seconds and you really need 8 parallel requests at all times. The auto adjustment won't allow for this and manual interaction is required to adjust the limits.

Another example is if the endpoint creation is much slower than expected, let's say it takes 10s instead of 2s. Then only one parallel endpoint request will be allowed as ParallelRequests will be shrunk down automatically.

Based on the metrics, it should be possible to tell immediately whether automatically adjusted limits are reasonable or not. We will be happy to optimize with you.

> My issue with the second approach is that the end user--the one who initiated the pod creation--typically won't have the ability to make cluster wide changes. Hence, my question is, what is the mode for interrogation and the passing of these failures and settings across the multiple components (end-user, k8s operator, cilium)?

Agreed. The user should not really see this. Similar to how a k8s user typically also doesn't see if the k8s apiserver returns 429. It does. Retries are done automatically. If too many retries are needed, the platform operator has to step in and resize the VM, adjust the limits, and so on. It will be the same with these limits. Observation should be done via the metrics so you see the mean waiting duration, if that gets out of hand, you can check why it is the case and determine whether to adjust the limits or fix resource constraints.